### ISSUE TITLE
Architecture

### ISSUE THREAD

**Issue by xrotwang** [cldf/cldfbench#2]
I'm trying to come to grips with the architecture of this thing. Clearly, `pyexibank` should be some sort of specialization of `cldfbench`. But how exactly should this look like?

Take the functionality to create a skeleton of a dataset (see #1): For lexibank, this command would need to use a specialized template for `README.md`. Now we could do this like it's done for the special MPG `clld` app template: `pylexibank` declares its special scaffold using `entry_points`, and the `cldfbench new-dataset` command takes a scaffold id as argument and looks up entry points.

For all cli mediated functionality this should work, i.e. `cldfbench` provides the cli infrastructure, and a way to register special behaviour for subcommands.

Possibly, this level of integration is enough, because most (all?) other special handling wil be implemented in the derived `Dataset` classes per dataset. And this code may simply import other utilities from `pylexibank`.

Does that sound reasonable?

**Comment by chrzyki** [comment]
Very much so! Strongly in favor!

**Comment by SimonGreenhill** [comment]
Makes sense, I think.

**Comment by xrotwang** [comment]
@SimonGreenhill @LinguList @chrzyki @tresoldi
Here's another architecture question: From my experience, the curation REPL wasn't really as useful as I hoped, and in particular made it even more complicated to take advantage of all the goodies we already get from the terminal shell. 

So I'd be inclined to drop this functionality in favour of a better defined command line interface. What do you think? Did any of you use (or even rely on) the command or dataset ID completion features? Or otherwise spend significant time in the REPL?

The cli I'm working on now has dataset discovery via "path to module" - so this again would help a lot with regular terminal automation, e.g. simple shell scripts à la "for p in `ls */lexibank_*.py` do" should just work.

**Comment by chrzyki** [comment]
I think leaving out the REPL (and also the dependencies it brings with it) is a good thing! As soon was `download`/`makecldf` were exposed directly through the CLI I never used the REPL again.

**Comment by SimonGreenhill** [comment]
strongly agree -- I only use the command line.

**Comment by xrotwang** [comment]
@SimonGreenhill @tresoldi @chrzyki @LinguList 
I wrote a small tutorial for the current `cldfbench` functionality, which outlines the basics: https://github.com/cldf/cldfbench/blob/master/doc/tutorial.md
Let me know what you think. It feels like a good start to me.

**Comment by LinguList** [comment]
I will try to have a look and implement my sound inventories data for Chinese dialects along these lines.

**Comment by xrotwang** [comment]
@LinguList yes, having a couple more fully implemented datasets should give us an idea about more generic functionality to go into `cldfbench`.

Note that `cldfbench` already comes with extensibility, so `pylexibank` would
- add its own dataset template, requiring more metadata
- provide additional sub-commands - to be run within `cldfbench`. E.g. `lexibank.makecldf` would not only require a `Glottolog` instance, but also `Concepticon`.

**Comment by SimonGreenhill** [comment]
Can we change “cldfbench_xxx.py” to just “cldf_xxx.py”

**Comment by xrotwang** [comment]
To make it shorter?
This would require adding another template variable when rendering the
dataset template. Right now it simply uses the package name.

Simon J Greenhill <notifications@github.com> schrieb am Fr., 11. Okt. 2019,
12:07:

> Can we change “cldfbench_xxx.py” to just “cldf_xxx.py”
>

**Comment by xrotwang** [comment]
Here's another architecture question: Currently in pylexibank the workflow commands `Dataset.cmd_*` accept arbitrary keyword arguments, and when called these keyword arguments are derived from an `argparse.Namespace` via `**vars(args)`. I'd find it more explicit if these methods accepted a single positional argument `args` instead - although there's still a bit of a disconnect because the subparser used to assemble this namespace would de defined in comand module somewhere else. But at least it would be clear that the arguments come from the command line.

**Comment by SimonGreenhill** [comment]
Re: cldf_* over cldfbench_*, mainly less typing, shorter file names and better branding, but this is probably just bikeshedding on my part
